File: vador/src/main/java/com/salesforce/vador/config/container/ContainerValidationConfigWith2Levels.java
/*******************************************************************************
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 ******************************************************************************/

package com.salesforce.vador.config.container;

import com.salesforce.vador.config.base.BaseContainerValidationConfig;
import de.cronn.reflection.util.TypedPropertyGetter;
import java.util.Collection;
import java.util.Set;
import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.Singular;
import lombok.Value;
import lombok.experimental.FieldDefaults;
import lombok.experimental.SuperBuilder;
import org.jetbrains.annotations.Nullable;

@Value
@FieldDefaults(level = AccessLevel.PACKAGE)
@EqualsAndHashCode(callSuper = true)
@SuperBuilder(buildMethodName = "prepare", builderMethodName = "toValidate", toBuilder = true)
public class ContainerValidationConfigWith2Levels<
        ContainerRootLevelValidatableT, ContainerLevel1ValidatableT, FailureT>
    extends BaseContainerValidationConfig<ContainerRootLevelValidatableT, FailureT> {

  @Singular
  Collection<
          TypedPropertyGetter<
              ContainerRootLevelValidatableT, @Nullable Collection<ContainerLevel1ValidatableT>>>
      withBatchMembers;

  ContainerValidationConfig<ContainerLevel1ValidatableT, FailureT> withScopeOf1LevelDeep;

  public Set<String> getFieldNamesForBatchLevel1(
      Class<ContainerLevel1ValidatableT> validatableClazz) {
    return ContainerValidationConfigEx.getFieldNamesForBatchLevel1Ex(this, validatableClazz);
  }

  public Set<String> getFieldNamesForBatchRootLevel(
      Class<ContainerRootLevelValidatableT> validatableClazz) {
    return ContainerValidationConfigEx.getFieldNamesForBatchEx(this, validatableClazz);
  }
}

File: vador/src/main/kotlin/com/salesforce/vador/config/container/ContainerValidationConfigWith2Levels.kt
/**
 * ****************************************************************************
 * Copyright (c) 2022, salesforce.com, inc. All rights reserved. SPDX-License-Identifier:
 * BSD-3-Clause For full license text, see the LICENSE file in the repo root or
 * https://opensource.org/licenses/BSD-3-Clause
 * ****************************************************************************
 */
package com.salesforce.vador.config.container

import com.salesforce.vador.config.base.BaseContainerValidationConfig
import com.salesforce.vador.immutables.ConfigStyle
import com.salesforce.vador.types.Validator
import com.salesforce.vador.types.ValidatorEtr
import de.cronn.reflection.util.TypedPropertyGetter
import org.immutables.value.Value

internal abstract class AbstractContainerValidationConfigWith2LevelsSupport<
  ContainerRootLevelValidatableT,
  ContainerLevel1ValidatableT,
  FailureT,
> : BaseContainerValidationConfig<ContainerRootLevelValidatableT, FailureT>() {

  abstract val withBatchMembers:
    Collection<
      TypedPropertyGetter<ContainerRootLevelValidatableT, Collection<ContainerLevel1ValidatableT>?>
    >
}

@ConfigStyle
@Value.Immutable(copy = false)
internal abstract class AbstractContainerValidationConfigWith2Levels<
  ContainerRootLevelValidatableT,
  ContainerLevel1ValidatableT,
  FailureT,
> :
  AbstractContainerValidationConfigWith2LevelsSupport<
    ContainerRootLevelValidatableT,
    ContainerLevel1ValidatableT,
    FailureT,
  >() {

  @get:Value.Auxiliary
  protected abstract val containerValidatorEtrs:
    List<ValidatorEtr<ContainerRootLevelValidatableT, FailureT>>

  @get:Value.Derived
  override val withContainerValidatorEtrs:
    Collection<ValidatorEtr<ContainerRootLevelValidatableT, FailureT>>
    get() = containerValidatorEtrs

  @get:Value.Auxiliary
  protected abstract val containerValidatorMappings:
    Map<Validator<in ContainerRootLevelValidatableT, FailureT>, FailureT>

  @get:Value.Derived
  override val withContainerValidator:
    Map<Validator<in ContainerRootLevelValidatableT, FailureT>, FailureT>
    get() = containerValidatorMappings

  @get:Value.Auxiliary
  protected abstract val batchMemberGetters:
    List<
      TypedPropertyGetter<ContainerRootLevelValidatableT, Collection<ContainerLevel1ValidatableT>?>
    >

  @get:Value.Derived
  override val withBatchMembers:
    Collection<
      TypedPropertyGetter<ContainerRootLevelValidatableT, Collection<ContainerLevel1ValidatableT>?>
    >
    get() = batchMemberGetters

  abstract val withScopeOf1LevelDeep:
    BaseSingleLevelContainerValidationConfig<ContainerLevel1ValidatableT, FailureT>

  @Value.NonAttribute
  fun getFieldNamesForBatchLevel1(
    validatableClazz: Class<ContainerLevel1ValidatableT>
  ): Set<String> = getFieldNamesForBatchLevel1Ex(validatableClazz)

  @Value.NonAttribute
  fun getFieldNamesForBatchRootLevel(
    validatableClazz: Class<ContainerRootLevelValidatableT>
  ): Set<String> = getFieldNamesForBatchEx(validatableClazz)

  abstract class Builder<ContainerRootLevelValidatableT, ContainerLevel1ValidatableT, FailureT> :
    ContainerValidationBuilderDsl<
      ContainerRootLevelValidatableT,
      FailureT,
      ContainerValidationConfigWith2Levels.Builder<
        ContainerRootLevelValidatableT,
        ContainerLevel1ValidatableT,
        FailureT,
      >,
    >,
    TwoLevelContainerBatchMembersBuilderDsl<
      ContainerRootLevelValidatableT,
      ContainerLevel1ValidatableT,
      ContainerValidationConfigWith2Levels.Builder<
        ContainerRootLevelValidatableT,
        ContainerLevel1ValidatableT,
        FailureT,
      >,
    >
}
